plugins/inx/server_utxo.go

```diff
@@ -0,0 +1,185 @@
+package inx
+
+import (
+	"context"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/gohornet/hornet/pkg/inx"
+	"github.com/gohornet/hornet/pkg/model/milestone"
+	"github.com/gohornet/hornet/pkg/model/utxo"
+	"github.com/iotaledger/hive.go/events"
+	"github.com/iotaledger/hive.go/workerpool"
+	iotago "github.com/iotaledger/iota.go/v3"
+)
+
+func (s *INXServer) ReadOutput(_ context.Context, id *inx.OutputId) (*inx.OutputResponse, error) {
+	deps.UTXOManager.ReadLockLedger()
+	defer deps.UTXOManager.ReadUnlockLedger()
+
+	ledgerIndex, err := deps.UTXOManager.ReadLedgerIndexWithoutLocking()
+	if err != nil {
+		return nil, err
+	}
+
+	outputID := id.Unwrap()
+
+	unspent, err := deps.UTXOManager.IsOutputIDUnspentWithoutLocking(outputID)
+	if err != nil {
+		return nil, err
+	}
+
+	if unspent {
+		output, err := deps.UTXOManager.ReadOutputByOutputID(outputID)
+		if err != nil {
+			return nil, err
+		}
+		ledgerOutput, err := inx.NewLedgerOutput(output)
+		if err != nil {
+			return nil, err
+		}
+		return &inx.OutputResponse{
+			LedgerIndex: uint32(ledgerIndex),
+			Payload: &inx.OutputResponse_Output{
+				Output: ledgerOutput,
+			},
+		}, nil
+	}
+
+	spent, err := deps.UTXOManager.ReadSpentForOutputIDWithoutLocking(outputID)
+	if err != nil {
+		return nil, err
+	}
+	ledgerSpent, err := inx.NewLedgerSpent(spent)
+	if err != nil {
+		return nil, err
+	}
+	return &inx.OutputResponse{
+		LedgerIndex: uint32(ledgerIndex),
+		Payload: &inx.OutputResponse_Spent{
+			Spent: ledgerSpent,
+		},
+	}, nil
+}
+
+func (s *INXServer) ReadUnspentOutputs(_ *inx.NoParams, srv inx.INX_ReadUnspentOutputsServer) error {
+	deps.UTXOManager.ReadLockLedger()
+	defer deps.UTXOManager.ReadUnlockLedger()
+
+	ledgerIndex, err := deps.UTXOManager.ReadLedgerIndexWithoutLocking()
+	if err != nil {
+		return err
+	}
+
+	var innerErr error
+	err = deps.UTXOManager.ForEachUnspentOutput(func(output *utxo.Output) bool {
+		ledgerOutput, err := inx.NewLedgerOutput(output)
+		if err != nil {
+			innerErr = err
+			return false
+		}
+		payload := &inx.UnspentOutput{
+			LedgerIndex: uint32(ledgerIndex),
+			Output:      ledgerOutput,
+		}
+		if err := srv.Send(payload); err != nil {
+			innerErr = err
+			return false
+		}
+		return true
+	})
+	if innerErr != nil {
+		return innerErr
+	}
+	return err
+}
+
+func (s *INXServer) ListenToLedgerUpdates(req *inx.LedgerUpdateRequest, srv inx.INX_ListenToLedgerUpdatesServer) error {
+	startIndex := milestone.Index(req.GetStartMilestoneIndex())
+	if startIndex > 0 {
+		// Stream all available milestone diffs first
+		pruningIndex := deps.Storage.SnapshotInfo().PruningIndex
+		if startIndex <= pruningIndex {
+			return status.Errorf(codes.InvalidArgument, "given startMilestoneIndex %d is older than the current pruningIndex %d", startIndex, pruningIndex)
+		}
+
+		deps.UTXOManager.ReadLockLedger()
```

Comment:
Pruning could interfere here, since you lock the ledger after checking the pruning index.
Do we have a pruning lock or something? Or is it sufficient to lock the ledger before?

Reply:
I am now locking the ledger before, but I think that manual pruning calls are not locking the UTXO ledger.